### ISSUE TITLE
Type-F OU model gives wrong likelihood when H is non-diagonal

### ISSUE THREAD

**Issue by hckiang** [venelin/PCMBase#6]
In the OU model, when H is non-diagonal (regardless of symmetric or not),

`modOU = PCM(PCMDefaultModelTypes()[["F"]], k = 2) `

and

`modOU = PCM("OU", k = 2) `

gives different result, and the former output a slightly wrong number, usually about 10% off from the correct likelihood.

Not sure if the former is "just the wrong way of calling the function", but the [vignette](https://cran.r-project.org/web/packages/PCMBase/vignettes/PCMBase.html) suggest so (implicitly, at least for me when I read it). If this is not a bug then I guess some clarification should be added into the vignette to prevent users making the same mistake.

**Comment by venelin** [comment]
Have you double-checked that not only the model type, but also the parameter values are the same?
Note that OU is not the same as type 'F', because type F only allows for matrices H with non-negative eigenvalues, while OU allows for negative eigenvalues.
If upon double checking you are still observing different likelihood values, please, provide a coding example. To save time, for the tree and the data, you can use some of the example objects coming with the package.

**Comment by hckiang** [comment]
I'm sure H has positive eigenvalues. Here is some stand-alone example code at my hand:

https://gist.github.com/hckiang/71ccb09b3d2ab3ec0c7067e1f5b803a0

**Comment by venelin** [comment]
Hello, 

I tried your code and I've slightly modified your functions so that they return a list of the log-likelihood value and the model:

```
pcmbase_oulik = function (tree, tipvals, H, theta, sig_x, rootval) {
  ## modOU = PCM(PCMDefaultModelTypes()[["F"]], k = length(rootval))
  modOU = PCM("OU", k = length(rootval))
  modOU$H[,,1] = H
  modOU$Theta[] = theta
  modOU$Sigma_x[,,1] = sig_x
  modOU$X0[] = rootval

  L = PCMLik(t(tipvals[tree$tip.label,,drop=F]), tree, modOU)
  #attr(L, "X0") = NULL
  list(L = L, model = modOU)
}

pcmbase_typeF_lik = function (tree, tipvals, H, theta, sig_x, rootval) {
  modOU = PCM(PCMDefaultModelTypes()[["F"]], k = length(rootval))
  modOU$H[,,1] = H
  modOU$Theta[] = theta
  modOU$Sigma_x[,,1] = sig_x
  modOU$X0[] = rootval

  L = PCMLik(t(tipvals[tree$tip.label,,drop=F]), tree, modOU)
  #attr(L, "X0") = NULL
  list(L = L, model = modOU)
}
```

Now, if you print the two models you will notice that the typeF model is _Transformable, while the type OU model is not: 

```
> pcmbase_typeF$model
Ornstein-Uhlenbeck model
S3 class: OU__Global_X0__Schur_WithNonNegativeDiagonal_Transformable_H__Theta__UpperTriangularWithDiagonal_WithNonNegativeDiagonal_Sigma_x__Omitted_Sigmae_x, OU, GaussianPCM, PCM, 
**_Transformable**; k=2; p=11; regimes: 1. Parameters/sub-models:
X0 (VectorParameter, _Global, numeric; trait values at the root):
[1] -1.2  2.0
H (MatrixParameter, _Schur, _WithNonNegativeDiagonal, _Transformable; adaptation rate matrix):
, , 1

     [,1] [,2]
[1,]  3.0  0.8
[2,]  0.8  2.0

Theta (VectorParameter, matrix; long-term optimum):
       1
[1,] 1.0
[2,] 1.3
Sigma_x (MatrixParameter, _UpperTriangularWithDiagonal, _WithNonNegativeDiagonal; factor of the unit-time variance rate):
, , 1

     [,1] [,2]
[1,]  1.4 -0.6
[2,]  0.0  1.4


> pcmbase_ou$model
Ornstein-Uhlenbeck model
S3 class: OU, GaussianPCM, PCM; k=2; p=14; regimes: 1. Parameters/sub-models:
X0 (VectorParameter, _Global, numeric; trait values at the root):
[1] -1.2  2.0
H (MatrixParameter; adaptation rate matrix):
, , 1

     [,1] [,2]
[1,]  3.0  0.8
[2,]  0.8  2.0

Theta (VectorParameter, matrix; long-term optimum):
       1
[1,] 1.0
[2,] 1.3
Sigma_x (MatrixParameter, _UpperTriangularWithDiagonal, _WithNonNegativeDiagonal; factor of the unit-time variance rate):
, , 1

     [,1] [,2]
[1,]  1.4 -0.6
[2,]  0.0  1.4

Sigmae_x (MatrixParameter, _UpperTriangularWithDiagonal, _WithNonNegativeDiagonal; factor of the non-heritable variance or the variance of the measurement error):
, , 1

     [,1] [,2]
[1,]    0    0
[2,]    0    0
```

Specifically, in the typeF model the matrix H is transformed via the so called Schur transformation before evaluating the log-likelihood. This transformation ensures that the eigenvalues of the transformed H are equal to the diagonal elements of H before transformation. This part of PCMBase is described in the [Parametrisation vignette](https://venelin.github.io/PCMBase/articles/PCMParam.html#transformation). 

To summarise, when you call PCMLik on pcmbase_typeF$model, first PCMApplyTransformation will be called on this on this model object and the value of the parameter H will change. I think this, explains the observed difference in the likelihood values and this behaviour is desired, so not a bug.  I'll close the issue now, feel free to open a new one if encounter other issues. 

Hope that helps, 
Venelin
